src/providers/vmware/amd64.rs

```diff
@@ -39,8 +57,152 @@ impl VmwareProvider {
     fn fetch_guestinfo(erpc: &mut vmw_backdoor::EnhancedChan, key: &str) -> Result<Option<String>> {
         let guestinfo = erpc
             .get_guestinfo(key.as_bytes())
-            .context("failed to retrieve network kargs")?
+            .context(format!("failed to retrieve guestinfo for {}", key))?
             .map(|bytes| String::from_utf8_lossy(&bytes).into_owned());
         Ok(guestinfo)
     }
+
+    pub fn parse_netplan_config(&self) -> Result<Option<netplan_types::NetplanConfig>> {
+        if let Some(metadata) = &self.guestinfo_metadata {
+            // To provide at least some sanity checking, parse the netplan config.
+            // First check if we have JSON userdata.
+            let netplan_config: netplan_types::NetplanConfig =
+                if let Ok(netplan_json_config) = serde_json::from_str(metadata) {
+                    netplan_json_config
+                } else {
+                    // The JSON parsing error is discarded,
+                    // and if YAML parsing fails as well, we only return this error.
+                    serde_yaml::from_str(metadata)?
+                };
+            // Also check the version to make it easier for the OS to say that
+            // this mechanism is supported or not.
+            if netplan_config.network.version != 2 {
+                bail!("only netplan version 2 supported");
+            }
+            Ok(Some(netplan_config))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_from_metadata(metadata: String) -> Result<Self> {
+        Ok(Self {
+            guestinfo_net_kargs: None,
+            guestinfo_metadata: Some(metadata),
+        })
+    }
+}
+
+fn parse_metadata(
+    guestinfo_metadata_encoding: Option<String>,
+    guestinfo_metadata_raw: Option<String>,
+) -> Result<Option<String>> {
+    let guestinfo_metadata =
+        if let Some(guestinfo_metadata_encoding_val) = guestinfo_metadata_encoding {
+            match (
+                guestinfo_metadata_encoding_val.as_str(),
+                guestinfo_metadata_raw,
+            ) {
+                ("base64", Some(guestinfo_metadata_raw_val)) => {
```

Comment:
Please note there are multiple, documented options for expressing the encoding:

* b64
* base64
* gz+b64
* gzip+base64

Reply:
Ok, added that